### DISCUSSION TITLE
How to config the specific implementation of IActivityExecutionMiddleware in Program.cs?

### DISCUSSION THREAD

**Discussion by coolshun** [elsa-workflows/elsa-core#5197] [Q&A]
How to config the specific implementation of IActivityExecutionMiddleware in Program.cs?The code bellow ,but "UseActivityExecutionMiddleware" not found.

`
services.AddElsa(elsa => elsa
    .UseActivityExecutionMiddleware<CustomActivityMiddleware>()
    // ... other configurations
);
`

**Comment by glime-ai** [comment]
To configure a specific implementation of `IActivityExecutionMiddleware` in `Program.cs`, you should first ensure that you have created a middleware class that implements the `IActivityExecutionMiddleware` interface. Once you have your custom middleware class, you can add it to the Elsa services configuration.

However, the method `UseActivityExecutionMiddleware` does not exist in the Elsa core library as a built-in extension method. Instead, you can add your custom middleware directly to the service collection using the `AddScoped` or `AddSingleton` method, depending on your middleware's required lifecycle.

Here's an example of how you might register your custom activity execution middleware:

```csharp
public class CustomActivityMiddleware : IActivityExecutionMiddleware
{
    public async Task ExecuteAsync(ActivityExecutionContext context, Func<Task> next)
    {
        // Your custom logic here

        // Call the next middleware in the pipeline
        await next();
    }
}

// In Program.cs
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddElsa(elsa =>
{
    // ... other Elsa configurations
});

// Register your custom middleware with the DI container
builder.Services.AddScoped<IActivityExecutionMiddleware, CustomActivityMiddleware>();

var app = builder.Build();

// ... configure the app

app.Run();
```

In this example, `CustomActivityMiddleware` is your custom middleware that implements `IActivityExecutionMiddleware`. You register it with the dependency injection container using `AddScoped`. This will ensure that your middleware is used during the execution of activities within Elsa workflows.

Please note that the exact way to integrate custom middleware into Elsa may vary depending on the version of Elsa you are using and the specifics of your application. If you are using a version of Elsa that supports a specific middleware registration method, you should refer to the official documentation or the source code for guidance.

**Comment by glime-ai** [comment]
To configure a specific implementation of `IActivityExecutionMiddleware` in `Program.cs` for Elsa 3.0, you should use the `UseMiddleware` extension method on the `IActivityExecutionPipelineBuilder` interface. However, based on the context provided, it seems that the method `UseActivityExecutionMiddleware` does not exist or is not found. This could be due to a missing using directive, an incorrect namespace, or the method not being available in the version of Elsa you are using.

Instead, you can try to define an extension method that adds your custom middleware to the activity execution pipeline. Here's an example based on the context provided:

```csharp
using Elsa.Workflows.Contracts;
using Elsa.Workflows.Pipelines.ActivityExecution;

namespace YourNamespace
{
    public static class CustomActivityMiddlewareExtensions
    {
        public static IActivityExecutionPipelineBuilder UseCustomActivityMiddleware(this IActivityExecutionPipelineBuilder pipelineBuilder)
        {
            return pipelineBuilder.UseMiddleware<CustomActivityMiddleware>();
        }
    }
}
```

And then in your `Program.cs`, you would use this extension method as part of the Elsa services configuration:

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddElsa(elsa =>
{
    // ... other configurations
    elsa.UsePipeline(pipeline => pipeline.UseCustomActivityMiddleware());
});

// ... rest of the code to build and run the application
```

Make sure that your `CustomActivityMiddleware` class implements the `IActivityExecutionMiddleware` interface and that you have the correct using directives for the namespaces where the interfaces and extension methods are defined.

If you continue to encounter issues, please check the Elsa documentation or the Elsa GitHub repository for the most up-to-date information on configuring middleware in Elsa 3.0.
